src/constraint_functions.jl

```diff
@@ -93,4 +91,400 @@ function EMB.constraints_level_aux(m, n::HydroStorage, 𝒯, 𝒫, modeltype::En
+            )
+        )
+    else
+        opex_var_discharge = @expression(m, [t_inv ∈ 𝒯ᴵⁿᵛ], 0)
+    end
+
+    # Create the constraint penalty constraint
+    constraints = constraint_data(n)
+    constraints_up = filter(has_penalty_up, constraints) # Max and schedule
+    constraints_down = filter(has_penalty_down, constraints) # Min and schedule
+
+    p = storage_resource(n)
+    if length(constraints_up) > 0
+        c_up = first(constraints_up)
+        penalty_up_var = @expression(m, [t_inv ∈ 𝒯ᴵⁿᵛ], sum(m[:rsv_penalty_up][n, t, p] *
+            penalty(c_up, t) * scale_op_sp(t_inv, t) for t ∈ t_inv if has_penalty(c_up, t)))
+    else
+        penalty_up_var = @expression(m, [t_inv ∈ 𝒯ᴵⁿᵛ], 0)
+    end
+
+    if length(constraints_down) > 0
+        c_down = first(constraints_down)
+        penalty_down_var = @expression(m, [t_inv ∈ 𝒯ᴵⁿᵛ], sum(m[:rsv_penalty_down][n, t, p] *
+            penalty(c_down, t) * scale_op_sp(t_inv, t) for t ∈ t_inv if has_penalty(c_down, t)))
+    else
+        penalty_down_var = @expression(m, [t_inv ∈ 𝒯ᴵⁿᵛ], 0)
+    end
+
+    # Create the overall constraint
+    @constraint(m, [t_inv ∈ 𝒯ᴵⁿᵛ],
+        m[:opex_var][n, t_inv] == opex_var_level[t_inv] + opex_var_charge[t_inv] +
+            opex_var_discharge[t_inv] + penalty_up_var[t_inv] + penalty_down_var[t_inv]
+    )
+end
+function EMB.constraints_opex_var(m, n::HydroUnit, 𝒯ᴵⁿᵛ, modeltype::EnergyModel)
+    constraints = constraint_data(n)
+
+    opex_var = @expression(m, [t_inv ∈ 𝒯ᴵⁿᵛ], sum(m[:cap_use][n, t] * EMB.opex_var(n, t) *
+        scale_op_sp(t_inv, t) for t ∈ t_inv))
+
+    penalty_up_var = Dict(t_inv => AffExpr(0) for t_inv ∈ 𝒯ᴵⁿᵛ)
+    penalty_down_var = Dict(t_inv => AffExpr(0) for t_inv ∈ 𝒯ᴵⁿᵛ)
+
+    penalty_up_var = @expression(m, [t_inv ∈ 𝒯ᴵⁿᵛ],
+        sum(penalty(c, t) * scale_op_sp(t_inv, t) * m[:gen_penalty_up][n, t, p]
+            for t ∈ t_inv
+            for p ∈ [water_resource(n), electricity_resource(n)]
+            for c in constraints
+            if has_penalty_up(c, t, p)
+        )
+    )
+    penalty_down_var = @expression(m, [t_inv ∈ 𝒯ᴵⁿᵛ],
+        sum(penalty(c, t) * scale_op_sp(t_inv, t) * m[:gen_penalty_down][n, t, p]
+            for t ∈ t_inv
+            for p ∈ [water_resource(n), electricity_resource(n)]
+            for c in constraints
+            if has_penalty_down(c, t, p)
+        )
+    )
+
+    @constraint(m, [t_inv ∈ 𝒯ᴵⁿᵛ],
+        m[:opex_var][n, t_inv] == opex_var[t_inv] + penalty_up_var[t_inv] +
+            penalty_down_var[t_inv]
+    )
+end
+
+"""
+    build_constraint(m, n::Union{HydroGate, HydroUnit}, c::Constraint, 𝒯::TimeStructure, p::ResourceCarrier)
+
+Create minimum/maximum/schedule discharge constraints for the generic `Node` type. The
+`Constraint{T}` can have types `T <: AbstractConstraintType` that defines the direction of
+the constraint.
+Penalty variables are included unless penalty value is not set or `Inf``.
+"""
+function build_constraint(
```

Comment:
We can streamline the construction of these constraints. That is not required for the current PR, but should be mentioned as an Issue.

Reply:
I have created an issue to ensure this will be resolved in the future:
https://github.com/EnergyModelsX/EnergyModelsRenewableProducers.jl/issues/36